WordPress/gutenberg
# Components: Add support for TypeScript unit tests

## What?


Following the exploratory work in #39341, this PR adds the necessary configuration in order to add unit tests written in TypeScript to the repo.

## Why?


Having unit tests in TypeScript is a way to ensure that we type-check a component from the point of view of a consumer (and not just internally).

This PR will allow us to refactor to TypeScript unit tests in the `@wordpress/components` package, in the context of the wider package refactor (https://github.com/WordPress/gutenberg/issues/35744).

## How?


- added a new TypeScript project specific for unit tests (`tsconfig.test.json`)
- added type defs for custom TypeScript matchers (`toMatchStyleDiffSnapshot`)
- edited `eslint` config to include TypeScript tests in the "development files" glob matcher
- refactored unit tests for the `<Text />` component to make sure that the above changes work for TypeScript unit tests

## Testing Instructions





- Project build
- All unit tests pass
- Running `npm run build && npx tsc -p packages/components/tsconfig.json` doesn't output any errors to console.
- Open `packages/components/src/text/test/index.tsx` in an IDE supporting TypeScript:
     - make sure that type defs and autocomplete is working as expected.
     - introduce a TypeScript error, make sure it gets flagged in your IDE and when running `npx tsc -p packages/components/tsconfig.json`
